File: Backend/app/routes/flight_groups.py
import asyncio
import logging
import re
from datetime import datetime

from fastapi import APIRouter

from app.config import settings
from app.services.aero_api import get_flight_realtime
from app.services.mongodb_service import get_database

router = APIRouter(tags=["FlightGroups"])
logger = logging.getLogger(__name__)


def _date_str(dt) -> str:
    if isinstance(dt, datetime):
        return dt.strftime("%Y-%m-%d")
    return ""


def _time_str(dt) -> str:
    if isinstance(dt, datetime):
        return dt.strftime("%H:%M")
    return ""


def _group_id(flight_type: str, flight_key: str, date_str: str) -> str:
    safe = re.sub(r"[^a-zA-Z0-9]", "", flight_key)
    safe_date = date_str.replace("-", "") or "nodate"
    return f"{flight_type}_{safe}_{safe_date}"


def _build_groups(bookings: list, flight_map: dict, flight_type: str) -> tuple[list, list]:
    """Group bookings by (flight_number, date) for the given flight_type.

    flight_type: "arrival" | "departure"
    Returns (groups_list, passengers_list).
    """
    groups: dict[str, dict] = {}
    passengers: list[dict] = []

    for booking in bookings:
        bid = str(booking["_id"])
        fd = flight_map.get(bid, {})
        section = fd.get(flight_type, {})

        flight_num  = (section.get("flight_number") or "").strip()
        flight_name = (section.get("flight_name")   or "").strip()

        if not flight_num and not flight_name:
            continue

        dt_key    = f"{flight_type}_datetime"
        dt_val    = section.get(dt_key)
        date_str  = _date_str(dt_val)
        time_str  = _time_str(dt_val)
        flight_key = flight_num or flight_name
        gid = _group_id(flight_type, flight_key, date_str)

        if gid not in groups:
            airport = (section.get("airport") or "").strip()
            groups[gid] = {
                "id":            gid,
                "flightNumber":  flight_num or flight_name,
                "airline":       flight_name or flight_num,
                "scheduledTime": time_str,
                "actualTime":    time_str,
                "date":          date_str,
                "terminal":      "",
                "type":          flight_type,
                "status":        "on_time",
                "origin":        airport if flight_type == "arrival"   else None,
                "destination":   airport if flight_type == "departure" else None,
                "passengerIds":  [],
            }

        groups[gid]["passengerIds"].append(bid)

        contact = booking.get("contact", {})
        full_name = f"{contact.get('first_name', '')} {contact.get('last_name', '')}".strip() or "Unknown"
        passengers.append({
            "id":                 bid,
            "name":               full_name,
            "phone":              contact.get("phone", ""),
            "mandal":             contact.get("mandal", ""),
            "passengerCount":     booking.get("passengers_count", 1),
            "destination":        "",
            "wheelchairRequired": bool(booking.get("stroller_required", False)),
            "carSeatRequired":    False,
            "flightGroup":        gid,
        })

    return list(groups.values()), passengers


async def _load_data():
    db = get_database()
    bookings = await db[settings.bookings_collection].find({}).to_list(None)
    if not bookings:
        return [], {}
    booking_ids = [b["_id"] for b in bookings]
    flight_docs = await db[settings.flight_details_collection].find(
        {"booking_id": {"$in": booking_ids}}
    ).to_list(None)
    flight_map = {str(fd["booking_id"]): fd for fd in flight_docs}
    return bookings, flight_map


async def _enrich_with_realtime(groups: list, flight_type: str) -> None:
    """Fetch AeroAPI status for each unique (flightNumber, date) and update groups in-place."""
    if not groups:
        return
    enrichments = await asyncio.gather(
        *[
            get_flight_realtime(g["flightNumber"], g["date"], flight_type, g["scheduledTime"])
            for g in groups
        ],
        return_exceptions=True,
    )
    for g, result in zip(groups, enrichments):
        if not isinstance(result, dict):
            continue
        g["status"] = result["status"]
        if result["actualTime"]:
            g["actualTime"] = result["actualTime"]
        if result["terminal"]:
            g["terminal"] = result["terminal"]


@router.get("/arrivals")
async def get_arrival_groups():
    bookings, flight_map = await _load_data()
    if not bookings:
        return {"groups": [], "passengers": []}
    groups, passengers = _build_groups(bookings, flight_map, "arrival")
    groups.sort(key=lambda g: (g["date"], g["scheduledTime"]))
    await _enrich_with_realtime(groups, "arrival")
    return {"groups": groups, "passengers": passengers}


@router.get("/departures")
async def get_departure_groups():
    bookings, flight_map = await _load_data()
    if not bookings:
        return {"groups": [], "passengers": []}
    groups, passengers = _build_groups(bookings, flight_map, "departure")
    groups.sort(key=lambda g: (g["date"], g["scheduledTime"]))
    await _enrich_with_realtime(groups, "departure")
    return {"groups": groups, "passengers": passengers}

File: Backend-js/src/routes/flight_groups.ts
import { Router } from 'express';
import { Booking, FlightDetails } from '../models/index.js';
import { ah } from '../util/asyncHandler.js';
import { getFlightRealtime } from '../services/aero_api.js';

function dateStr(dt: unknown): string {
  if (dt instanceof Date) {
    const pad = (n: number) => String(n).padStart(2, '0');
    return `${dt.getUTCFullYear()}-${pad(dt.getUTCMonth() + 1)}-${pad(dt.getUTCDate())}`;
  }
  return '';
}

function timeStr(dt: unknown): string {
  if (dt instanceof Date) {
    const pad = (n: number) => String(n).padStart(2, '0');
    return `${pad(dt.getUTCHours())}:${pad(dt.getUTCMinutes())}`;
  }
  return '';
}

function groupId(flightType: string, flightKey: string, date: string): string {
  const safe = flightKey.replace(/[^a-zA-Z0-9]/g, '');
  const safeDate = date.replace(/-/g, '') || 'nodate';
  return `${flightType}_${safe}_${safeDate}`;
}

interface Group {
  id: string;
  flightNumber: string;
  airline: string;
  scheduledTime: string;
  actualTime: string;
  date: string;
  terminal: string;
  type: string;
  status: string;
  origin: string | null;
  destination: string | null;
  passengerIds: string[];
}

function buildGroups(bookings: any[], flightMap: Map<string, any>, flightType: 'arrival' | 'departure') {
  const groups = new Map<string, Group>();
  const passengers: any[] = [];

  for (const booking of bookings) {
    const bid = String(booking._id);
    const fd = flightMap.get(bid) ?? {};
    const section = fd[flightType] ?? {};

    const flightNum = String(section.flight_number ?? '').trim();
    const flightName = String(section.flight_name ?? '').trim();
    if (!flightNum && !flightName) continue;

    const dtKey = `${flightType}_datetime`;
    const dtVal = section[dtKey];
    const date = dateStr(dtVal);
    const time = timeStr(dtVal);
    const flightKey = flightNum || flightName;
    const gid = groupId(flightType, flightKey, date);

    if (!groups.has(gid)) {
      const airport = String(section.airport ?? '').trim();
      groups.set(gid, {
        id: gid,
        flightNumber: flightNum || flightName,
        airline: flightName || flightNum,
        scheduledTime: time,
        actualTime: time,
        date,
        terminal: '',
        type: flightType,
        status: 'on_time',
        origin: flightType === 'arrival' ? airport : null,
        destination: flightType === 'departure' ? airport : null,
        passengerIds: [],
      });
    }
    groups.get(gid)!.passengerIds.push(bid);

    const contact = booking.contact ?? {};
    const fullName =
      `${contact.first_name ?? ''} ${contact.last_name ?? ''}`.trim() || 'Unknown';
    passengers.push({
      id: bid,
      name: fullName,
      phone: contact.phone ?? '',
      email: contact.email ?? '',
      mandal: contact.mandal ?? '',
      passengerCount: booking.passengers_count ?? 1,
      destination: '',
      wheelchairRequired: Boolean(booking.stroller_required),
      carSeatRequired: false,
      flightGroup: gid,
      trackingToken: booking.tracking_token ?? null,
    });
  }

  return { groups: [...groups.values()], passengers };
}

async function loadData() {
  const bookings = await Booking.find({}).lean();
  if (bookings.length === 0) return { bookings: [], flightMap: new Map<string, any>() };
  const bookingIds = bookings.map((b: any) => b._id);
  const flightDocs = await FlightDetails.find({ booking_id: { $in: bookingIds } }).lean();
  const flightMap = new Map<string, any>(flightDocs.map((fd: any) => [String(fd.booking_id), fd]));
  return { bookings, flightMap };
}

async function enrichWithRealtime(groups: Group[], flightType: string): Promise<void> {
  if (groups.length === 0) return;
  const results = await Promise.all(
    groups.map((g) =>
      getFlightRealtime(g.flightNumber, g.date, flightType, g.scheduledTime).catch(() => null)
    )
  );
  for (let i = 0; i < groups.length; i++) {
    const result = results[i];
    if (!result) continue;
    groups[i].status = result.status;
    if (result.actualTime) groups[i].actualTime = result.actualTime;
    if (result.terminal) groups[i].terminal = result.terminal;
  }
}

export const flightGroupsRouter = Router();

flightGroupsRouter.get(
  '/arrivals',
  ah(async (_req, res) => {
    const { bookings, flightMap } = await loadData();
    if (bookings.length === 0) {
      res.json({ groups: [], passengers: [] });
      return;
    }
    const { groups, passengers } = buildGroups(bookings, flightMap, 'arrival');
    groups.sort((a, b) => (a.date + a.scheduledTime).localeCompare(b.date + b.scheduledTime));
    await enrichWithRealtime(groups, 'arrival');
    res.json({ groups, passengers });
  })
);

flightGroupsRouter.get(
  '/departures',
  ah(async (_req, res) => {
    const { bookings, flightMap } = await loadData();
    if (bookings.length === 0) {
      res.json({ groups: [], passengers: [] });
      return;
    }
    const { groups, passengers } = buildGroups(bookings, flightMap, 'departure');
    groups.sort((a, b) => (a.date + a.scheduledTime).localeCompare(b.date + b.scheduledTime));
    await enrichWithRealtime(groups, 'departure');
    res.json({ groups, passengers });
  })
);
